Add SAML addon.

very-ape opened the pull request:
I'm running a set of services for friends and family and using Keycloak for account management. All of the services I run use SAML to create a smooth, integrated experience via single sign-on and single logout.

This addon replaces the signup and login mechanisms with SAML-based SSO (and SLO).

Commit:
Add SAML addon.

Commit:
Log SAML errors in saml_sso_reply

MrPetovan commented on saml/saml.php:
```diff
@@ -0,0 +1,429 @@
+    Hook::register('login_hook', 'addon/saml/saml.php', 'saml_sso_initiate');
+    Hook::register('logging_out', 'addon/saml/saml.php', 'saml_slo_initiate');
+    Hook::register('head', 'addon/saml/saml.php', 'saml_head');
+    Hook::register('footer', 'addon/saml/saml.php', 'saml_footer');
```
Portability: This code is correct, but we've started using `__FILE__` instead to improve code reuse between addons and prevent rare issues with moved files.
```suggestion
    Hook::register('login_hook', __FILE__, 'saml_sso_initiate');
    Hook::register('logging_out', __FILE__, 'saml_slo_initiate');
    Hook::register('head', __FILE__, 'saml_head');
    Hook::register('footer', __FILE__, 'saml_footer');
```

MrPetovan commented on saml/saml.php:
```diff
@@ -0,0 +1,429 @@
+        saml_input(
+            'settings_statement',
+            'Settings statement',
+            'A statement on the settings page explaining where the user should go to change their e-mail and password. HTML allowed.'
```
Please reconsider injecting a config value as a HTML string in the page. I would recommend enabling BBCode, the HMTL output is sanitized according to pretty strict standards. The method to convert BBCode to HTML is `\Friendica\Content\Text\BBCode::convert()`.

MrPetovan commented on saml/saml.php:
```diff
@@ -0,0 +1,429 @@
+            $key,
+            DI::l10n()->t($label),
+            DI::config()->get('saml', $key),
+            DI::l10n()->t($description),
```
Unfortunately, these dynamic translation strings won't be picked up by our translation string collector script and as a result won't be available for our volunteer translators to translate in Transifex. You have to explicitly wrap each translation string with a `DI::l10n()->t()` call.

MrPetovan commented on saml/saml.php:
```diff
@@ -0,0 +1,429 @@
+
+function saml_module($a) {}
+
+function saml_init($a) {
```
Code standards: We use a very slight variation of the PSR-2 coding standards that mandates a new line before function opening braces among other things, please check the rest of the basic rules: https://friendica.mrpetovan.com/help/Developers-Intro#Coding+standards

MrPetovan commented on saml/saml.php:
```diff
@@ -0,0 +1,429 @@
+                \OneLogin\Saml2\Error::METADATA_SP_INVALID
+            );
+        }
+    } catch (Exception $e) {
+	Logger::error($e->getMessage());
```
Coding standards: Mixed indentation style, please use only tabs.

MrPetovan commented on saml/saml.php:
```diff
@@ -0,0 +1,429 @@
+}
+
+function saml_sso_initiate(&$a, &$b) {
+    if (!saml_is_configured()) return;
```
Coding standards: Please add braces to all conditions.

MrPetovan commented on saml/templates/admin.tpl:
```diff
@@ -0,0 +1,36 @@
+{{include file="field_textarea.tpl" field=$sp_cert}}
+{{include file="field_textarea.tpl" field=$idp_cert}}
+
+<!--
+<script type="text/javascript">
+
+function select_all()
+
+{
+
+var text_val = document.getElementById('id_sp_cert');
+
+text_val.focus();
+
+text_val.select();
+
+document.execCommand("Copy");
+
+}
+
+</script>
+
+<div class="field textarea">
+	<label for="id_sp_cert">SP certificate</label>
+	<textarea id="id_sp_cert" aria-describedby="sp_cert_tip" onClick="select_all();">{{$sp_cert}}</textarea>
+	<span class="field_help" role="tooltip" id="sp_cert_tip">The certificate this addon will use to authenticate with your identity provider. Click it to copy it.</span>
+</div>
+-->
```
As a general rule, we don't include code comments in commits since we are using a versioning tool. Why does it need to be added to this file?

Commit:
Use __FILE__ instead of hardcoded string in hook registration.

Commit:
Switch from HTML to BBCode for 'Basic Settings' notice.

Commit:
Only pass string literals to l10n. Only string literals passed directly get picked up by the translation service.

Commit:
Remove overlooked HTML comment.

Commit:
Replace spaces with tabs.

Commit:
Move function declaration braces to meet PSR-2 standard.

Commit:
Remove superfluous 'use' statement. The SAML library comes with its own loader.

Commit:
Apply PHPCS fixes.

Commit:
Remove debug endpoint.

Commit:
Break up long line.

Commit:
Prevent execution of SAML functions and show or log a message when the addon isn't configured.

very-ape commented:
Thanks for the quick turnaround on that code review! I believe I've addressed all the problems you noted and have also added some friendly messages for when the addon triggers before it's configured.

MrPetovan commented on saml/saml.php:
```diff
@@ -0,0 +1,455 @@
+
+require_once(__DIR__ . '/vendor/autoload.php');
+
+define("PW_LEN", 32); // number of characters to use for random passwords
```
Coding standards: Please use single quotes unless double quotes are needed for variable interpolation/special characters (like `\n`).

MrPetovan commented on saml/saml.php:
```diff
@@ -0,0 +1,455 @@
+			$label,
+			DI::config()->get('saml', $key),
+			$description,
+		true, // all the fields are required
```
Coding standards: Wrong indentation.

MrPetovan commented on saml/saml.php:
```diff
@@ -0,0 +1,455 @@
+function saml_addon_admin_post(&$a)
+{
+	$safeset = function ($key) {
+		$val = (!empty($_POST[$key]) ? Strings::escapeTags(trim($_POST[$key])) : '');
```
It used to be the norm, but `Strings::escapeTags` is now deprecated for processing user input. Instead, we rely on Smarty's default HTML escaping on display.

MrPetovan commented on saml/saml.php:
```diff
@@ -0,0 +1,455 @@
+	$safeset('sp_cert');
+	$safeset('idp_cert');
+
+	// Not using safeset here since settings_statement is *meant* to include HTML tags.
```
With the switch to BBCode, this comment and the exception are now obsolete.

MrPetovan commented on saml/saml.php:
```diff
@@ -0,0 +1,455 @@
+			'nickname' => $username,
+			'email'	=> $email,
+			'password' => base64_encode($bytes), // should be at least PW_LEN long
+		'verified' => true
```
Coding standards: Wrong indentation.

MrPetovan commented on saml/saml.php:
```diff
@@ -0,0 +1,455 @@
+	   'security' => array (
+		   'wantXMLValidation' => false,
+
+	   // Indicates whether the <samlp:AuthnRequest> messages sent by this SP
+	   // will be signed.  [Metadata of the SP will offer this info]
+	   'authnRequestsSigned' => true,
+
+	   // Indicates whether the <samlp:logoutRequest> messages sent by this SP
+	   // will be signed.
+	   'logoutRequestSigned' => true,
+
+	   // Indicates whether the <samlp:logoutResponse> messages sent by this SP
+	   // will be signed.
+	   'logoutResponseSigned' => true,
+
+	   /* Sign the Metadata */
+	   'signMetadata' => true,
```
Coding standards: Wrong indentation.

MrPetovan commented on saml/saml.php:
```diff
@@ -0,0 +1,455 @@
+	   // will be signed.
+	   'logoutResponseSigned' => true,
+
+	   /* Sign the Metadata */
```
Coding standards: Multi-line comment used for single-line comment.

MrPetovan reviewed (changes requested):
Thank you for acting on the feedback, I have some more for you!

Commit:
Replace double quotes with single quotes where applicable.

Commit:
Fix wrong indentation! D:

Commit:
Remove the deprecated Strings::escapeTags, as we now rely on Smarty to catch HTML tags.

Commit:
Replace array() with square bracket syntax.

Commit:
Add whitespace around comments in saml_settings to ease reading.

very-ape commented:
Okay, I _think_ I got everything this time.

Commit:
Set a reasonable version number!

MrPetovan reviewed (approved):
Looks good to me, thanks!

tobiasd commented:
With #1113 merged, the translation of the SAML addon is now in the pipe for the translation teams
